var/spack/repos/builtin/packages/py-mock/package.py

```diff
@@ -12,14 +12,17 @@ class PyMock(PythonPackage):
     they have been used."""
 
     homepage = "https://github.com/testing-cabal/mock"
-    pypi = "mock/mock-3.0.5.tar.gz"
+    pypi = "mock/mock-4.0.3.tar.gz"
 
+    version('4.0.3', sha256='7d3fbbde18228f4ff2f1f119a45cdffa458b4c0dee32eb4d2bb2f82554bac7bc')
     version('3.0.5', sha256='83657d894c90d5681d62155c82bda9c1187827525880eda8ff5df4ec813437c3')
     version('2.0.0', sha256='b158b6df76edd239b8208d481dc46b6afd45a846b7812ff0ce58971cf5bc8bba')
     version('1.3.0', sha256='1e247dbecc6ce057299eb7ee019ad68314bb93152e81d9a6110d35f4d5eca0f6')
 
     depends_on('python@2.7:2.8,3.4:', type=('build', 'run'))
```

Comment:
If we do this, I think it would be nicer to write a script which generates the required depend_on's from the pypi archives.

Instead of manually checking all versions, a script would be far better and it could run as pack command, even in the pipeline, without human labour to check all the dependencies and versions.

All I wanted was to ensure that another package (bump of py-argcomplete) which's PR I reviewed breaks no dependents, but then that build-of-all-the dependents uncovered that py-gsutil didn't build because it has a depends_on py-monotonic@1.4: but no version new enough was in spack, so I bumped py-monotonic, and the found that py-mock@2.0.0 which it needs didn't build because of missing py-pbr, and in the build-out log, it also said:
```
WARNING: The wheel package is not available.
```
So I add both, and hope that's it, but then (and so on so on so so on) I've to stop somewhere.

We've to automate that fixing if we really want to go beyond just fixing that a simple build works, I did this just for fun and not for work, so I'd have to stop otherwise if we can't limit the depths we take or you could check what it all needs.

Ultimatively, it should be an automatic scripted check for the majority of things if possible.

Reply:
Thanks, limiting the old versions to when='@:2.0.0')` works:
```py
depends_on('python@2.7:2.8,3.4:', type=('build', 'run'), when='@:2.0.0')`
```

Reply:
@adamjstewart Was there maybe once work on such an automatic dependency checker?

For a start, it such checker could run as part of  `lib/spack/spack/build_systems/python.py`:

It could be similar to like we check that `autoconf`, `automake` and `libtool` are installed before we try to run `autoreconf` in `lib/spack/spack/build_systems/autotools.py`

I'm sure it would not catch all cases, but it could issue warnings when a package listed in requirements.txt isn't installed in the correct version. Such check should be in setuptools itself as well or not?

Reply:
I've never tried to write an automatic dependency checker, and I don't think it's possible in general. Dependencies can be listed in `setup.cfg`, `pyproject.toml` (Python doesn't come with a toml parser), or `setup.py` (impossible to parse raw Python). I've written pretty extensive [documentation](https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html) on this to help new contributors determine a package's dependencies. Also, we can't really check at build time since Spack needs to know the full list of dependencies ahead of time.

Reply:
`pip-compile` from https://github.com/jazzband/pip-tools can generate and update `requirements.txt` files,
and `pip-sync` from there can update a venv according to the the generated requirements.txt file. That could be a start:

Like pip-sync, but instead generate spack-warnings when the provided or generated requirements,txt misses things, and when python specs are installed which are not required.

Reply:
We probably can't vendor these deps in Spack just for Python packages, but if you want to try to write a script to auto-generate Spack packages that would be great. This has been done for R already where the ecosystem is much more standardized. My only concern is that a script like that would pull in style/test dependencies that are not actually required to build the package. It also wouldn't have any concept of build vs build/run deps.